### PR TITLE
[Utilities] Created utility to get an outline of all module features

### PR DESCRIPTION
**Pull request by MrMCake**
# Description

- Created the utility that returns an array/markdown table of the most important module features (e.g. RBAC, child-modules, Private Endpoints, etc.)

# Example
## Command
`> Get-ModulesFeatureOutline -returnMarkdown -onlyTopLevel`

## Output
| # | Module | RBAC | Locks | Tags | Diag | PE | PIP | # children | # lines |
| - | - | - | - | - | - | - | - | - | - |
| 1 | MS.AAD<p>DomainServices | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 226 |
| 2 | MS.AnalysisServices<p>servers | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 141 |
| 3 | MS.ApiManagement<p>service | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:11, L2:3] | 411 |
| 4 | MS.AppConfiguration<p>configurationStores | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:1] | 202 |
| 5 | MS.Authorization<p>locks |  |  |  |  |  |  | [L1:2] | 57 |
| 6 | MS.Authorization<p>policyAssignments |  |  |  |  |  |  | [L1:3] | 130 |
| 7 | MS.Authorization<p>policyDefinitions |  |  |  |  |  |  | [L1:2] | 84 |
| 8 | MS.Authorization<p>policyExemptions |  |  |  |  |  |  | [L1:3] | 104 |
| 9 | MS.Authorization<p>policySetDefinitions |  |  |  |  |  |  | [L1:2] | 75 |
| 10 | MS.Authorization<p>roleAssignments |  |  |  |  |  |  | [L1:3] | 104 |
| 11 | MS.Authorization<p>roleDefinitions |  |  |  |  |  |  | [L1:3] | 91 |
| 12 | MS.Automation<p>automationAccounts | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:6] | 352 |
| 13 | MS.Batch<p>batchAccounts |  | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 184 |
| 14 | MS.CognitiveServices<p>accounts | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | 246 |
| 15 | MS.Compute<p>availabilitySets | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 81 |
| 16 | MS.Compute<p>diskEncryptionSets | :white_check_mark: |  | :white_check_mark: |  |  |  |  | 103 |
| 17 | MS.Compute<p>disks | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 168 |
| 18 | MS.Compute<p>galleries | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | [L1:1] | 100 |
| 19 | MS.Compute<p>images | :white_check_mark: |  | :white_check_mark: |  |  |  |  | 70 |
| 20 | MS.Compute<p>proximityPlacementGroups | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 69 |
| 21 | MS.Compute<p>virtualMachines | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:1] | 575 |
| 22 | MS.Compute<p>virtualMachineScaleSets | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:1] | 570 |
| 23 | MS.Consumption<p>budgets |  |  |  |  |  |  |  | 90 |
| 24 | MS.ContainerInstance<p>containerGroups |  | :white_check_mark: | :white_check_mark: |  |  |  |  | 113 |
| 25 | MS.ContainerRegistry<p>registries | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:2] | 305 |
| 26 | MS.ContainerService<p>managedClusters | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:1] | 493 |
| 27 | MS.Databricks<p>workspaces | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 139 |
| 28 | MS.DataFactory<p>factories | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:2] | 226 |
| 29 | MS.DataProtection<p>backupVaults | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | [L1:1] | 105 |
| 30 | MS.DesktopVirtualization<p>applicationgroups | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:1] | 145 |
| 31 | MS.DesktopVirtualization<p>hostpools | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 177 |
| 32 | MS.DesktopVirtualization<p>scalingplans | :white_check_mark: |  | :white_check_mark: | :white_check_mark: |  |  |  | 136 |
| 33 | MS.DesktopVirtualization<p>workspaces | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 117 |
| 34 | MS.DocumentDB<p>databaseAccounts | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:2, L2:2] | 243 |
| 35 | MS.EventGrid<p>systemTopics | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 137 |
| 36 | MS.EventGrid<p>topics | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | 149 |
| 37 | MS.EventHub<p>namespaces | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:4, L2:2] | 279 |
| 38 | MS.HealthBot<p>healthBots | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 66 |
| 39 | MS.Insights<p>actionGroups | :white_check_mark: |  | :white_check_mark: |  |  |  |  | 83 |
| 40 | MS.Insights<p>activityLogAlerts | :white_check_mark: |  | :white_check_mark: |  |  |  |  | 72 |
| 41 | MS.Insights<p>components | :white_check_mark: |  | :white_check_mark: |  |  |  |  | 97 |
| 42 | MS.Insights<p>diagnosticSettings |  |  |  | :white_check_mark: |  |  |  | 79 |
| 43 | MS.Insights<p>metricAlerts | :white_check_mark: |  | :white_check_mark: |  |  |  |  | 120 |
| 44 | MS.Insights<p>privateLinkScopes | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | :white_check_mark: |  | [L1:1] | 95 |
| 45 | MS.Insights<p>scheduledQueryRules | :white_check_mark: |  | :white_check_mark: |  |  |  |  | 104 |
| 46 | MS.KeyVault<p>vaults | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:3] | 282 |
| 47 | MS.KubernetesConfiguration<p>extensions |  |  |  |  |  |  |  | 63 |
| 48 | MS.KubernetesConfiguration<p>fluxConfigurations |  |  |  |  |  |  |  | 67 |
| 49 | MS.Logic<p>workflows | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 192 |
| 50 | MS.MachineLearningServices<p>workspaces | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:1] | 259 |
| 51 | MS.ManagedIdentity<p>userAssignedIdentities | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 62 |
| 52 | MS.ManagedServices<p>registrationDefinitions |  |  |  |  |  |  |  | 60 |
| 53 | MS.Management<p>managementGroups |  |  |  |  |  |  |  | 40 |
| 54 | MS.NetApp<p>netAppAccounts | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | [L1:1, L2:1] | 104 |
| 55 | MS.Network<p>applicationGateways | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 306 |
| 56 | MS.Network<p>applicationSecurityGroups | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 61 |
| 57 | MS.Network<p>azureFirewalls | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | :white_check_mark: |  | 260 |
| 58 | MS.Network<p>bastionHosts | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | :white_check_mark: |  | 194 |
| 59 | MS.Network<p>connections |  | :white_check_mark: | :white_check_mark: |  |  |  |  | 107 |
| 60 | MS.Network<p>ddosProtectionPlans | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 62 |
| 61 | MS.Network<p>expressRouteCircuits | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 186 |
| 62 | MS.Network<p>firewallPolicies |  |  | :white_check_mark: |  |  |  | [L1:1] | 152 |
| 63 | MS.Network<p>frontDoors | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 148 |
| 64 | MS.Network<p>ipGroups | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 66 |
| 65 | MS.Network<p>loadBalancers | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:2] | 236 |
| 66 | MS.Network<p>localNetworkGateways | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 86 |
| 67 | MS.Network<p>natGateways | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 182 |
| 68 | MS.Network<p>networkInterfaces | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 142 |
| 69 | MS.Network<p>networkSecurityGroups | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:1] | 150 |
| 70 | MS.Network<p>networkWatchers | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | [L1:2] | 95 |
| 71 | MS.Network<p>privateDnsZones | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | [L1:9] | 187 |
| 72 | MS.Network<p>privateEndpoints | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | [L1:1] | 97 |
| 73 | MS.Network<p>publicIPAddresses | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 166 |
| 74 | MS.Network<p>publicIPPrefixes | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 72 |
| 75 | MS.Network<p>routeTables | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 68 |
| 76 | MS.Network<p>trafficmanagerprofiles | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 162 |
| 77 | MS.Network<p>virtualHubs |  | :white_check_mark: | :white_check_mark: |  |  |  | [L1:2] | 148 |
| 78 | MS.Network<p>virtualNetworkGateways | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 359 |
| 79 | MS.Network<p>virtualNetworks | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:2] | 228 |
| 80 | MS.Network<p>virtualWans | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 78 |
| 81 | MS.Network<p>vpnGateways |  | :white_check_mark: | :white_check_mark: |  |  |  | [L1:2] | 104 |
| 82 | MS.Network<p>vpnSites | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 88 |
| 83 | MS.OperationalInsights<p>workspaces | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:4] | 241 |
| 84 | MS.OperationsManagement<p>solutions |  |  |  |  |  |  |  | 50 |
| 85 | MS.RecoveryServices<p>vaults | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:6, L2:2, L3:1] | 268 |
| 86 | MS.Resources<p>deploymentScripts |  | :white_check_mark: | :white_check_mark: |  |  |  |  | 111 |
| 87 | MS.Resources<p>resourceGroups | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 62 |
| 88 | MS.Resources<p>tags |  |  | :white_check_mark: |  |  |  | [L1:2] | 51 |
| 89 | MS.Security<p>azureSecurityCenter |  |  |  |  |  |  |  | 217 |
| 90 | MS.ServiceBus<p>namespaces | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:7, L2:2] | 307 |
| 91 | MS.ServiceFabric<p>clusters | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | [L1:1] | 279 |
| 92 | MS.Sql<p>managedInstances | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  | [L1:6, L2:2] | 324 |
| 93 | MS.Sql<p>servers | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | :white_check_mark: |  | [L1:4] | 204 |
| 94 | MS.Storage<p>storageAccounts | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:5, L2:4, L3:1] | 345 |
| 95 | MS.Synapse<p>privateLinkHubs | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | :white_check_mark: |  |  | 85 |
| 96 | MS.VirtualMachineImages<p>imageTemplates | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 163 |
| 97 | MS.Web<p>connections | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  |  | 86 |
| 98 | MS.Web<p>hostingEnvironments | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 162 |
| 99 | MS.Web<p>serverfarms | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  |  |  | 155 |
| 100 | MS.Web<p>sites | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | [L1:2] | 248 |
| 101 | MS.Web<p>staticSites | :white_check_mark: | :white_check_mark: | :white_check_mark: |  | :white_check_mark: |  |  | 144 |
| = | | 78 | 76 | 86 | 46 | 17 | 2 | 137 | 16534 |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

### CODE REVIEW

File: docs/wiki/The library - Module overview.md
Comment (eriqua) [inline docs/wiki/The library - Module overview.md:13]: I'd suggest either to expand column names or better a legend to explain e.g. what EP means

File: .github/workflows/platform.updateReadMe.yml
Comment (eriqua) [inline .github/workflows/platform.updateReadMe.yml:34]: Should the script path 'utilities/tools/platform' be an environment variable?

File: .azuredevops/platformPipelines/platform.updateReadMe.yml
Comment (eriqua) [reply .azuredevops/platformPipelines/platform.updateReadMe.yml:39]: Should the pipelines be triggered on changes to scripts in the 'utilities/tools/platform' folder?

File: docs/wiki/The library - Module overview.md
Comment (MrMCake) [reply docs/wiki/The library - Module overview.md:13]: I did have bigger names, but Mate asked me to shorten them. A legend seems to be a good compromise - will add

File: .azuredevops/platformPipelines/platform.updateReadMe.yml
Comment (MrMCake) [reply .azuredevops/platformPipelines/platform.updateReadMe.yml:39]: Sure we can add a trigger. Note however, this will also affect a script that is out of scope of this PR - the `Set-ReadMeModuleTable` script

Comment (MrMCake) [reply .azuredevops/platformPipelines/platform.updateReadMe.yml:39]: Regarding the environment variable I'm still not sold on the idea :D But sure I could set it somewhere. The only problem with the path in the environment would be that I have to choose whether to use `/` or `\` in the path. 'Join-Path' does that for me, based on the used OS.
Thoughts?

File: .github/workflows/platform.updateReadMe.yml
Comment (MrMCake) [reply .github/workflows/platform.updateReadMe.yml:34]: See comment above

File: docs/wiki/The library - Module overview.md
Comment (MrMCake) [reply docs/wiki/The library - Module overview.md:13]: added

File: .azuredevops/platformPipelines/platform.updateReadMe.yml
Comment (MrMCake) [reply .azuredevops/platformPipelines/platform.updateReadMe.yml]: For the time being I added them, though I'm still not sure. Also, how far do we want to push this? For example, we also have the snippet:

```PowerShell
$functionInput = @{
   MarkdownFilePath = Join-Path '$(System.DefaultWorkingDirectory)' 'docs' 'wiki' 'The library - Module overview.md'
   ModuleFolderPath = Join-Path '$(System.DefaultWorkingDirectory)' 'modules'
}
```

Should these then also be `env` values? Depending on the file the list 'may' get rather long.

File: docs/wiki/The library - Module overview.md
Comment (eriqua) [inline docs/wiki/The library - Module overview.md:128]:
```suggestion
| `PE` | Whether the module can deploy _Private Endpoints_. |
```

File: docs/wiki/The library - Module overview.md
Comment (eriqua) [inline docs/wiki/The library - Module overview.md:129]:
```suggestion
| `PIP` | Whether the module can deploy a _Public IP_ as a secondary resource. |
```

File: docs/wiki/The library - Module overview.md
Comment (eriqua) [inline docs/wiki/The library - Module overview.md:131]: Are those the number of lines of the deploy.bicep? Then I'd suggest something like
```suggestion
| `# lines` | The number of lines in the module Bicep deployment file. |
```

File: .azuredevops/platformPipelines/platform.updateReadMe.yml
Comment (eriqua) [reply .azuredevops/platformPipelines/platform.updateReadMe.yml]: I agree, let's discuss a consistent way forward offline as it impacts the whole repo. I just would like to avoid hardcoded values as much as possible, but I agree we should probably end up with a reasonable compromise. For this PR I'm fine either way.

File: docs/wiki/The library - Module overview.md
Comment (MrMCake) [inline docs/wiki/The library - Module overview.md:14]:
```suggestion
| # | Module | RBAC | Locks | Tags | Diag | EP | PIP | # children | # lines |
```

File: utilities/tools/Get-ModulesFeatureOutline.ps1
Comment (MrMCake) [inline utilities/tools/Get-ModulesFeatureOutline.ps1:136]:
```suggestion
            'PIP'  = $supportsPipDeployment
```

File: utilities/tools/Get-ModulesFeatureOutline.ps1
Comment (MrMCake) [inline utilities/tools/Get-ModulesFeatureOutline.ps1:136]:
```suggestion
            'PIP'          = $supportsPipDeployment
```

File: utilities/tools/Get-ModulesFeatureOutline.ps1
Comment (MrMCake) [inline utilities/tools/Get-ModulesFeatureOutline.ps1:136]:
```suggestion
            'PIP'        = $supportsPipDeployment
```

File: utilities/tools/Get-ModulesFeatureOutline.ps1
Comment (eriqua) [inline utilities/tools/Get-ModulesFeatureOutline.ps1:185]:
```suggestion
                    'PIP'  = $_.'PIP'
```

File: utilities/tools/Get-ModulesFeatureOutline.ps1
Comment (eriqua) [inline utilities/tools/Get-ModulesFeatureOutline.ps1:46]: This is something used in other scripts as well. I don't have a strong preference, but I wonder what's the advantage of this default value compared to the fact that we are forcing the script to be in a specific location in order to retrieve the modules folder + we are hardcoding the modules folder name as well. 
This function is anyway called by the pipeline passing parameter values to it, so the default values are never used and risk to get stale overtime if we for instance ever restructure the utilities folder. Food for thoughts. Anyway, I wouldn't block the PR on that since the same is done extensively throughout PS functions.

Comment (MrMCake) [reply utilities/tools/Get-ModulesFeatureOutline.ps1:46]: I actually ran the script most of the time locally. The docs update was an afterthought. The default value is by the way not enforcing the location of the script but merely makes it easier to use if the script is not moved. I don't really know what see any disadvantage here

Comment (eriqua) [reply utilities/tools/Get-ModulesFeatureOutline.ps1:46]: As mentioned, this is not blocking so let's discuss it offline. The main disadvantage I see is the risk of getting the default value stale. I'm sure you tested that locally in implementation phase. I'm not sure this would be often the case after merge.

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**
